drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,815 @@
+	synopGMAC_disable_interrupt_all(gmacdev);
+
+	/* Reset MAC */
+	synopGMAC_reset(gmacdev);
+	gmacdev->Intf = NUMAKER_GMAC_INTF;
+	synopGMAC_read_version(gmacdev);
+
+	/* Check for Phy initialization */
+	synopGMAC_set_mdc_clk_div(gmacdev, GmiiCsrClk5);
+	gmacdev->ClockDivMdc = synopGMAC_get_mdc_clk_div(gmacdev);
+
+	/* Reset PHY */
+	status = reset_phy(gmacdev);
+
+	/* Set up the tx and rx descriptor queue/ring */
+	synopGMAC_setup_tx_desc_queue(gmacdev, TRANSMIT_DESC_SIZE, RINGMODE);
+	synopGMAC_init_tx_desc_base(gmacdev);
+
+	synopGMAC_setup_rx_desc_queue(gmacdev, RECEIVE_DESC_SIZE, RINGMODE);
+	synopGMAC_init_rx_desc_base(gmacdev);
+
+	/* Initialize the dma interface */
+	synopGMAC_dma_bus_mode_init(gmacdev,
+				    DmaBurstLength32 | DmaDescriptorSkip0 | DmaDescriptor8Words);
+	synopGMAC_dma_control_init(gmacdev,
+				   DmaStoreAndForward | DmaTxSecondFrame | DmaRxThreshCtrl128);
+
+	/* Initialize the mac interface */
+	synopGMAC_mac_init(gmacdev);
+	synopGMAC_promisc_enable(gmacdev);
+
+	/* This enables the pause control in Full duplex mode of operation */
+	synopGMAC_pause_control(gmacdev);
+
+#if defined(NU_USING_HW_CHECKSUM)
+	/*IPC Checksum offloading is enabled for this driver. Should only be used if
+	 * Full Ip checksumm offload engine is configured in the hardware
+	 */
+	offload_needed = 1;
+
+	/* Enable the offload engine in the receive path */
+	synopGMAC_enable_rx_chksum_offload(gmacdev);
+
+	/* Default configuration, DMA drops the packets if error in encapsulated ethernet payload */
+	synopGMAC_rx_tcpip_chksum_drop_enable(gmacdev);
+#endif
+
+	for (i = 0; i < RECEIVE_DESC_SIZE; i++) {
+		skb = &rx_buf[NUMAKER_GMAC_INTF][i];
+		synopGMAC_set_rx_qptr(gmacdev, (u32)((u64)(skb->data) & NUMAKER_MASK_32),
+				      sizeof(skb->data), (u32)((u64)skb & NUMAKER_MASK_32));
+	}
+
+	for (i = 0; i < TRANSMIT_DESC_SIZE; i++) {
+		skb = &tx_buf[NUMAKER_GMAC_INTF][i];
+		synopGMAC_set_tx_qptr(gmacdev, (u32)((u64)(skb->data) & NUMAKER_MASK_32),
+				      sizeof(skb->data), (u32)((u64)skb & NUMAKER_MASK_32),
+				      offload_needed, 0);
+	}
+
+	synopGMAC_set_mac_address(NUMAKER_GMAC_INTF, mac_addr);
+
+	synopGMAC_clear_interrupt(gmacdev);
+
+	return status;
+}
+
+static int m_numaker_gmacdev_get_rx_buf(synopGMACdevice *gmacdev, uint16_t *len, uint8_t **buf)
+{
+	DmaDesc *rxdesc = gmacdev->RxBusyDesc;
+
+	LOG_DBG("---Start---");
+	if (synopGMAC_is_desc_owned_by_dma(rxdesc)) {
+		return -EIO;
+	}
+	if (synopGMAC_is_desc_empty(rxdesc)) {
+		return -ENODATA;
+	}
+
+	*len = synop_handle_received_data(NUMAKER_GMAC_INTF, buf);
+	if (*len <= 0) {
+		synopGMAC_enable_interrupt(gmacdev, DmaIntEnable);
+		return -ENOSPC; /* No available RX frame */
+	}
+
+	/* length of payload should be <= 1514 */
+	if (*len > (NU_ETH_MAX_FLEN - 4)) {
+		LOG_DBG("unexpected long packet length=%d, buf=0x%x\r", *len, (uint32_t)*buf);
+		*len = 0; /* Skip this unexpected long packet */
+	}
+	LOG_DBG("---End---\r");
```

Comment:
Note that you should avoid putting `\n` or `\r` to log statements as they typically will make output funny looking

Reply:
Yes, it's unnecesary in LOG_XX.
I fixed it.